### PR TITLE
Add Travis build

### PR DESCRIPTION
**Pull request by gliptak**
Signed-off-by: Gábor Lipták <gliptak@gmail.com>

WIP https://travis-ci.com/gliptak/book/builds/101868078

Would you have some pointers on how to correct

```
Unknown LaTeX class `per-file-class'
Makefile:21: recipe for target 'Crypto101.tex' failed
```

https://github.com/crypto101/book/issues/263

### CODE REVIEW

**Comment by lvh** [comment]
Thanks for working on this!

per-file-class is defined at the start of Crypto101.org in a pile of elisp:

```elisp
  (require 'cl) ;; The find function is in the cl library.
  (unless (find "per-file-class" org-latex-classes :key 'car :test 'equal)
    (add-to-list 'org-latex-classes
                 '("per-file-class"
                   "\\documentclass{memoir}"
                   ("\\part{%s}" . "\\part{%s}")
                   ("\\chapter{%s}" . "\\chapter*{%s}")
                   ("\\section{%s}" . "\\section*{%s}")
                   ("\\subsection{%s}" . "\\subsection*{%s}")
                   ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
                   ("\\paragraph{%s}" . "\\paragraph*{%s}")
                   ("\\subparagraph{%s}" . "\\subparagraph*{%s}"))))
  (add-to-list 'org-export-smart-quotes-alist
               '("en"
                 (opening-double-quote :utf-8 "“" :html "&ldquo;" :latex "\\enquote{" :texinfo "``")
                 (closing-double-quote :utf-8 "”" :html "&rdquo;" :latex "}" :texinfo "''")
                 (opening-single-quote :utf-8 "‘" :html "&lsquo;" :latex "\\enquote*{" :texinfo "`")
                 (closing-single-quote :utf-8 "’" :html "&rsquo;" :latex "}" :texinfo "'")
                 (apostrophe :utf-8 "’" :html "&rsquo;")))

  ;; Header.tex explicitly loads all necessary packages.
  (make-local-variable 'org-latex-default-packages-alist)
  (setf org-latex-default-packages-alist nil)
```

It effectively just sets the document class. At some point org-mode stopped running elisp for batch compilation. One approach might be to move that code in a separate elisp file, load that elisp file as part of the `-Q --batch` emacs call, and, if possible, fix the org-mode code loading?

**Comment by gliptak** [comment]
@lvh thank you for the pointers. I have limited familiarity with elips, so I would like help separating the elisp file and I would make the Travis updates needed right after.

**Comment by gliptak** [comment]
@lvh I tried to separate the elisp code unsuccessfully. Latex doesn't resolve when separated  ```Symbol's value as variable is void: org-latex-classes```

**Comment by lvh** [comment]
(Hey, I just wanted to ack this: been sick last few days and I'm catching up on backlog)

**Comment by gliptak** [comment]
no worries. i hope you can guide with elisp

File: org2tex [hunk 1]
```diff
@@ -2,18 +2,15 @@
...
-        --visit="$1" --funcall org-latex-export-to-latex
-else
-    echo "No org-mode directory found."
-fi
+emacs -Q --batch \
+	--eval "(require 'package)" \
+	--eval "(add-to-list 'package-archives '(\"org\" . \"https://orgmode.org/elpa/\"))" \
+	--eval "(package-initialize t)" \
+	--eval "(package-refresh-contents)" \
+	--eval "(package-install 'org-plus-contrib)"
```
Comment (lvh) [inline org2tex]: I understand why this is necessary but this is also potentially destructive for folks who already had org installed some other way :)

Consider, perhaps, moving this section into its own script that you only run for Travis?

**Comment by lvh** [comment]
See master; but you'll probably have to merge those org2tex conflicts manually.

FWIW here's what I had to install on fedora-29 to get the build to work:

```
dnf install texlive-glossaries texlive-minted texlive-wrapfig texlive-collection-metapost texlive-memoir texlive-adjustbox texlive-blindtext
```

File: org2tex (hunk 1 above)
Comment (gliptak) [reply org2tex]: @lvh After the build is green, I'm open to merge this with you original code checking for ```org```.

**Comment by gliptak** [comment]
@lvh would you have a (draft) ```Dockerfile``` available (or a list of commands to be "dockerized")? I could retrofit that https://docs.travis-ci.com/user/docker/
The  builds are failing on ```org```/```emacs```, so installing additional packages might have limited impact.

**Comment by lvh** [comment]
I'm only seeing builds from 10 days ago here: https://travis-ci.com/gliptak/book/builds and it looks like they're all still failing on the `per-file-class` thing I fixed last night?

(it also occurs to me now that `book` was not a good name for this repo)

**Comment by gliptak** [comment]
@lvh I didn't see your latest commit. I forced a build at https://travis-ci.com/gliptak/book/builds/102017131 but it failed with the same error (I will resolve merge conflict later today and rerun the build again)

PS yes ```book``` as a repo name is very generic :)

**Comment by gliptak** [comment]
Progress. After rebase,  https://travis-ci.com/gliptak/book/builds/103126553

```
! LaTeX Error: Option clash for package hyperref.
See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.13 \hypersetup
              {
? 
```

**Comment by gliptak** [comment]
@lvh do you recall seeing this above error in your setup?

**Comment by lvh** [comment]
Nope but the most obvious answer is that \hyperref is invoked more than once, and once without options. Can you try to exfiltrate the .tex file so we can debug it separately?

(My best guess is that it's a different version of emacs/org doing something wonky. Perhaps try building on Xenial or even macOS (since homebrew has new packages)? https://docs.travis-ci.com/user/reference/overview/)

**Comment by gliptak** [comment]
It is already ```xenial```

https://api.travis-ci.com/v3/job/182907714/log.txt

```
cat Crypto101.tex
% Created 2019-03-06 Wed 23:05
\documentclass[11pt,ebook,table,dvipsnames,oneside,openany]{memoir}
\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{fixltx2e}
\usepackage{graphicx}
\usepackage{longtable}
\usepackage{float}
\usepackage{wrapfig}
\usepackage{rotating}
\usepackage[normalem]{ulem}
\usepackage{amsmath}
\usepackage{textcomp}
\usepackage{marvosym}
\usepackage{wasysym}
\usepackage{amssymb}
\usepackage{hyperref}
\tolerance=1000
\input{Header}
\author{lvh}
\date{\today}
\title{Crypto 101}
\hypersetup{
  pdfkeywords={},
  pdfsubject={An introduction to cryptography for programmers},
  pdfcreator={Emacs 24.5.1 (Org mode 8.2.10)}}
\begin{document}

\maketitle
\tableofcontents
```

**Comment by gliptak** [comment]
yes, ```Fedora 29``` where it is at

https://travis-ci.com/gliptak/book/builds/103467771

**Comment by lvh** [comment]
Wait, seriously? I mean whatever I'll take it :)

**Comment by gliptak** [comment]
@lvh please review

after commit, I will submit a PR setting up ```releases``` https://docs.travis-ci.com/user/deployment/releases/

**Comment by gliptak** [comment]
@lvh any other changes you would like to see?

Please signup at https://travis-ci.com/ and enable Travis for your account.

**Comment by lvh** [comment]
Hey sorry, problem has just been that I've been busy at work, I already have a Travis account :)

**Comment by lvh** [comment]
1 more useful change: I think it would be better if Travis ran from a Dockerfile instead of just bash -cing something -- although you could also just get the bash -c source out of travis.yml. Part of the goal here is that _other people_ can also reliably build. However, this looks great already and I don't want perfect to be the enemy of good so I'm just going to merge this as is :)

Thank you _so much_ for working on this!

**Comment by lvh** [comment]
Wait, apparently travis-ci.org vs travis.ci.com has drastically changed. I thought it used to be the commercial offering vs the public one?

**Comment by lvh** [comment]
Can confirm: builds are working :) Thanks again!

**Pull request merged**